meetings/working-groups/collection-literals/LDM-questions-2023-08-15.md

```diff
@@ -0,0 +1,223 @@
+There are two primary directions we can go here:
+
+1. Analyze how the variable is use that to inform the scenario. Specifically, we could use flow analysis to determine what scope a variable should have.  For example:
+
+    ```c#
+    void M()
+    {
+        // Local scoped, s does not escape.
+        Span<int> s = [x, y, z];
+        foreach (var v in s)
+            Console.WriteLine(v);
+    }
+    ```
+
+    ```c#
+    ReadOnlySpan<int> M()
+    {
+        // Global scoped, since 's' escapes out of 'M'.
+        Span<int> s = [x, y, z];
+        UseSpan(s);
+        return s;
+    }
+    ```
+
+    Pros: Users can write spans + collection-expressions in a very simple fashion, and have the code "just work".
+
+    Cons: Unclear about what's actually happening.  Minor changes could lead to very different outcomes. Analysis is likely also very non trivial (on the order of major flow analysis work).  Scoping constraints would have to flow backwards from escape points back to the origination point.  For example:
+
+    ```c#
+    ReadOnlySpan<int> M()
+    {
+        Span<int> s1 = [x, y, z];
+        Span<int> s2 = X(s1);
+        return s2;
+    }
+
+    Span<T> X<T>(Span<T> s) => s;
+    ```
+
+    Here, 's2' must have global scope in order to escape 'M'.  But that then must flow that constraint into the invocation of 'X'.  This will then have to flow the constraint into 's1' to finally choose that s1 is globally scoped.  Complexity rises greatly with all flow-analysis constructs. 
+
+2. Decide on the scope directly from the declaration, instead of depending on how it is used.
+
+    - Option A: Ref-struct local has global scope by default.  To have local-scope, add the `scoped` keyword:
+
+        ```c#
+        void M()
+        {
+            // 's' has 'global scope' but does not escape.  Will allocate.  If they want stack-allocation, add `scoped`
+            Span<int> s = [x, y, z];
+            foreach (var v in s)
+                Console.WriteLine(v);
+        }
+        ```
+
+        Ideally compiler tells the user to make these variables `scoped` when it can be.
+
+        Pros: User gets the scoping they explicitly request.  When scope can be narrower, they are informed.
+
+        Cons: Analysis is similarly complex due to requisite reverse flow analysis.  Also, WG believes the default position of users is to want a span + collection-expression to have local scope and be stack allocated.  However, to get that, users will have to write:
+
+        ```c#
+        void M()
+        {
+            // 's' has 'local scope'
+            scoped Span<int> s = [x, y, z];
+            foreach (var v in s)
+                Console.WriteLine(v);
+        }
+        ```
+
+        This feels punishing to the common/default case that users will want.  Importantly, it means that if a user wants to switch off from `stackalloc` to collection-expressions they get less pleasant code. 
+
+    - Option B: Ref-struct local always has local scope.
```

Comment:
Should this be *collection expression* instead of just locals (for return values, method arguments, etc.)?
```suggestion
    - Option B: Ref struct collection expression always has local scope.
```